Source/Provers/SMTLib/SMTLibInteractiveTheoremProver.cs

```diff
@@ -514,62 +517,68 @@ private async Task<Model> GetErrorModel(CancellationToken cancellationToken)
       }
 
       SendThisVC("(get-model)");
-      Process.Ping();
-      Model theModel = null;
+
+      Model model = await ReadOutputAsync<Model>(
+        null,
+        (resp, theModel) => {
+          if (theModel != null)
+          {
+            HandleProverError("Expecting only one model but got many");
+          }
+
+          return ParseErrorModel(resp);
+        },
+        cancellationToken);
+
+      return model;
+    }
+    
+    /// <summary>
+    /// Sends a ping, and until the result is a pong,
+    /// passes intermediate results to collector
+    /// Returns the final result returned by collector, or the default value
+    /// </summary>
+    private async Task<T> ReadOutputAsync<T>(T defaultValue, Func<SExpr, T, T> collector, CancellationToken cancellationToken) {
+      Process.Ping1();
+
+      T result = defaultValue;
+
       while (true)
       {
         var resp = await Process.GetProverResponse().WaitAsync(cancellationToken);
-        if (resp == null || Process.IsPong(resp))
+        if (resp == null || Process.IsPong1(resp))
         {
           break;
         }
 
-        if (theModel != null)
-        {
-          HandleProverError("Expecting only one model but got many");
-        }
-
-        theModel = ParseErrorModel(resp);
+        result = collector(resp, result);
       }
 
-      return theModel;
+      return result;
     }
 
     private async Task<Outcome> GetResponse(CancellationToken cancellationToken)
     {
-      var result = Outcome.Undetermined;
       var wasUnknown = false;
 
-      Process.Ping();
-
-      while (true)
-      {
-        var resp = await Process.GetProverResponse().WaitAsync(cancellationToken);
-        if (resp == null || Process.IsPong(resp))
-        {
-          break;
-        }
-
-        result = ParseOutcome(resp, out wasUnknown);
-      }
+      var result = await ReadOutputAsync(
```

Comment:
nice generalization but maybe we don't need to pass in a lambda; could we just return the SExpr instead?

Reply:
Maybe a lot of sexprs, then?